### DISCUSSION TITLE
Migrating a source plugin to v4: documentation code examples seem incomplete

### DISCUSSION THREAD

**Discussion by cdanny358** [gatsbyjs/gatsby#34052] [Help]
In this [source plugin migration document](https://www.gatsbyjs.com/docs/reference/release-notes/migrating-source-plugin-from-v3-to-v4/), section 2 talks about data mutations happening during `sourceNodes` or `onCreateNode`. However, the example code for The New Way has this line:

`for (const { url } of remoteImages) {`

`remoteImages` isn't defined anywhere, and nothing like it appears in The Old Way. How is this variable supposed to be generated?

**Comment by LekoArts** [comment]
You'd get the `remoteImages` from the `// code to fetch data` portion. So you call an API, get a list of URLs and then you iterate over that to download the file and create the node.

**Reply by cdanny358** [reply]
So now, rather than the remote images automatically appearing during `createResolvers`, we have to manually fetch them, at a point in the process where `graphql` isn't available, and you have to use either an external graphql client, or resort to a REST API with axios, as `fetch` doesn't work inside `gatsby-node.js`? I'm not sure I see how this is progress.

**Reply by LekoArts** [reply]
In `createResolvers` you also manually fetched them with the helper function `createRemoteFileNode`, it only abstracted the fetching logic away. `fetch` not being available in Node.js is a node limitation, not a Gatsby one.

You have two options to migrate:

- https://www.gatsbyjs.com/docs/reference/release-notes/migrating-source-plugin-from-v3-to-v4/#2-data-mutations-need-to-happen-during-sourcenodes-or-oncreatenode
- Source your image in `onCreateNode`: https://www.gatsbyjs.com/docs/how-to/images-and-media/preprocessing-external-images/

In order for e.g. DSG to work correctly we had to make this change.

**Reply by cdanny358** [reply]
Perhaps I should be a bit more specific about the problems I'm running into: I use Directus as my database of choice, which has a graphql interface, a REST interface, and a dedicated SDK. To source items from Directus, I use `gatsby-source-graphql`, then add a `createResolvers` call to generate remote file nodes for all the images used by the program. If I try to move the file generation code to `onCreateNode`, I run into the problem that almost all the nodes returned by `gatsby-source-graphql` are of type `SitePage`, with single nodes of types `SiteBuildMetadata` and `GraphQLSource` at the head. None of these nodes contain any information about the remote files contained within. It seems like in Gatsby V4 I would be forced to create a Directus source plugin from scratch, using either the REST API or the dedicated SDK, which is a huge amount of work compared to the present situation. Am I missing something obvious?
